qa/tasks/cephfs/test_admin.py

```diff
@@ -2011,3 +2044,131 @@ def test_fs_authorize(self):
                 args=(f'fs authorize {self.fs.name} {self.CLIENT_NAME} / '
                       f'{wrong_perm}'), retval=self.EXPECTED_ERRNO,
                 errmsgs=self.EXPECTED_ERRMSG)
+
+
+class TestFSFail(TestAdminCommands):
+
+    MDSS_REQUIRED = 2
+    CLIENTS_REQUIRED = 1
+
+    def test_with_health_warn_oversize_cache(self):
+        health_warn = 'MDS_CACHE_OVERSIZED'
+        self.config_set('mds', 'mds_cache_memory_limit', '1K')
+        self.config_set('mds', 'mds_health_cache_threshold', '1.00000')
+
+        active_mds_id = self.fs.get_active_names()[0]
+        self.mount_a.open_n_background(".", 400)
+        self.wait_till_health_warn(health_warn, active_mds_id)
+
+        # actual testing begins now.
+        # test that "fs fail" fails without confirmation flag.
+        errmsg = 'mds_cache_oversized'
+        self.negtest_ceph_cmd(args=f'fs fail {self.fs.name}',
+                              retval=1, errmsgs=errmsg)
+        # test that "fs fail" passes with confirmation flag.
+        self.run_ceph_cmd(f'fs fail {self.fs.name} --yes-i-really-mean-it')
+
+    def test_with_health_warn_trim(self):
+        health_warn = 'MDS_TRIM'
+        active_mds_id = self.fs.get_active_names()[0]
+        self.mount_a.open_n_background(".", 400)
```

Comment:
Hang on - opening 400 files would result in MDS_TRIM warning? Am I reading this right that the config_set() done in the previous test will affect this test (and therefore we expect the trim warning)?

Reply:
Yes, opening 400 files would result in MDS_TRIM for a brief period. Originally it was 4000 files but I've reduced it to 400.

Reply:
Ran a couple more times, very very brief period here is 4 seconds.

Reply:
Is that with vstart_runner? I doubt we get a trim warning with the default cache sizes...

Reply:
Yes. There's no other way so fast to run tests with teuthology,

Reply:
I've updated the code. These config options are being set for every test now, so that we don't see such issues anymore.

Reply:
There are more reliable ways to do this and I think several are available. Use configs like `mds_debug_subtrees` (to force an `ESubtreeMap` for every event, `mds_log_warn_factor`, and `mds_log_max_segments` to get the MDS behind on trimming.

Play around with it and see.

Reply:
Current method to generate one of the health warning suffices while running these tests with vstart_runner.py. If they suffice with teuthology too, is it okay if I don't implement one of the methods above?

Reply:
I think it will often break by not throwing TRIM warnings. Neither Venky nor I can explain at first look why opening 400 files with a minisculel MDS cache would cause trim warnings. That should indicate this is a heinous hack that needs a new approach. :)

Reply:
> There are more reliable ways to do this and I think several are available. Use configs like `mds_debug_subtrees` (to force an `ESubtreeMap` for every event, `mds_log_warn_factor`, and `mds_log_max_segments` to get the MDS behind on trimming.
> 
> Play around with it and see.

Agree with @batrick here. For this we have hit the TRIM warning for many many time and this sounds more reliable approach to trigger the TRIM warnings.

Reply:
From the debug logs each `EUpdate openc` event will take 1K+ spaces:

```
2024-04-18T20:40:17.964+0000 7fa3d2238700  5 mds.1.log _submit_thread 8687941033580~1966 : EUpdate openc [metablob 0x2000277f5a0, 2 dirs]
2024-04-18T20:40:20.489+0000 7fa3d2238700  5 mds.1.log _submit_thread 8687941172472~1966 : EUpdate openc [metablob 0x2000277f5a0, 2 dirs]
2024-04-18T20:40:20.509+0000 7fa3d2238700  5 mds.1.log _submit_thread 8687941180085~1966 : EUpdate openc [metablob 0x2000277f5a0, 2 dirs]
2024-04-18T20:40:20.671+0000 7fa3d2238700  5 mds.1.log _submit_thread 8687941193940~1966 : EUpdate openc [metablob 0x200025eafe9, 2 dirs]
2024-04-18T20:40:20.846+0000 7fa3d2238700  5 mds.1.log _submit_thread 8687941201245~1964 : EUpdate openc [metablob 0x200025eafe9, 2 dirs]
2024-04-18T20:40:21.422+0000 7fa3d2238700  5 mds.1.log _submit_thread 8687941218460~1972 : EUpdate openc [metablob 0x200025eafe9, 2 dirs]
2024-04-18T20:40:21.516+0000 7fa3d2238700  5 mds.1.log _submit_thread 8687941224702~1968 : EUpdate openc [metablob 0x200025eafe9, 2 dirs]
2024-04-18T20:40:21.732+0000 7fa3d2238700  5 mds.1.log _submit_thread 8687941226690~5426 : EUpdate openc [metablob 0x1000866c4ca, 6 dirs]
2024-04-18T20:40:22.321+0000 7fa3d2238700  5 mds.1.log _submit_thread 8687941242965~1966 : EUpdate openc [metablob 0x2000277f5a0, 2 dirs]
2024-04-18T20:40:22.333+0000 7fa3d2238700  5 mds.1.log _submit_thread 8687941249955~1966 : EUpdate openc [metablob 0x2000277f5a0, 2 dirs]
2024-04-18T20:40:23.262+0000 7fa3d2238700  5 mds.1.log _submit_thread 8687941271836~1966 : EUpdate openc [metablob 0x200025eafe9, 2 dirs]
2024-04-18T20:40:23.439+0000 7fa3d2238700  5 mds.1.log _submit_thread 8687941296112~1964 : EUpdate openc [metablob 0x200025eafe9, 2 dirs]
2024-04-18T20:40:23.924+0000 7fa3d2238700  5 mds.1.log _submit_thread 8687941325931~1977 : EUpdate openc [metablob 0x2000288c274, 2 dirs]
2024-04-18T20:40:24.006+0000 7fa3d2238700  5 mds.1.log _submit_thread 8687941332925~1972 : EUpdate openc [metablob 0x200025eafe9, 2 dirs]
2024-04-18T20:40:24.098+0000 7fa3d2238700  5 mds.1.log _submit_thread 8687941336919~1968 : EUpdate openc [metablob 0x200025eafe9, 2 dirs]
2024-04-18T20:40:24.309+0000 7fa3d2238700  5 mds.1.log _submit_thread 8687941340909~5425 : EUpdate openc [metablob 0x1000866c4ca, 6 dirs]
```